Wasm.simd Negation, Splat, ReplaceLane + tests

Krovatkin opened the pull request:
This PR includes:
* Negation op  + tests
* Splat op + tests 
* ReplaceLane op + tests

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/microsoft/chakracore/3237)
<!-- Reviewable:end -->

Krovatkin commented:
@Cellule @MikeHolman please ignore this PR for now until I fix any possible issues w/ x-plat.

fyi @arunetm

Krovatkin commented on test/wasm.simd/replaceLaneTests.js:
```diff
@@ -0,0 +1,67 @@
+    let typeIndex = Math.log2(len) - 2;
+    let arr = arrays [typeIndex];
+
+    if (len == 32) { //special case for floats
```
we could change the indexing to a dict w/ an array view per each type where the type comes from a test name if this version is too ugly :-)

Krovatkin commented:
@Cellule could you please take a look when you have a moment? The tests seem to have cleared

Krovatkin commented:
@Cellule hate to bug you that often; could you please take a look when you have a moment?

Cellule commented:
Reviewed 15 of 15 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 1286 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3237#-KnuCpaRG4KU5hoglIEA:-KnuCpaSxjhXktwbeOO7:b2expv6) ([raw file](https://github.com/microsoft/chakracore/blob/cd79abbef5b7157181ea884af981da990bdaff7b/lib/WasmReader/WasmByteCodeGenerator.cpp#L1286)):*
> ```cpp
> EmitInfo WasmBytecodeGenerator::EnregisterIndex() 
> {
>     const uint offset = GetReader()->m_currentNode.lane.index;
> ```

I think I would put the code in `CheckLaneIndex` in here that way if you register an index you have to check the index is not out of range.
Also, could we rename this function to something like `EmitLaneIndex(Js::OpCode requestingOp)`

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 1303 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3237#-KnuDgEgunjxxElIemhi:-KnuDgEgunjxxElIemhj:b-9n323v) ([raw file](https://github.com/microsoft/chakracore/blob/cd79abbef5b7157181ea884af981da990bdaff7b/lib/WasmReader/WasmByteCodeGenerator.cpp#L1303)):*
> ```cpp
>     }
> 
>     EmitInfo simdArg = PopEvalStack();
> ```

use `PopEvalStack(WasmTypes::M128, _u("simd argument type mismatch"))` instead of checking the type manually after    

---

*[test/wasm.simd/replaceLaneTests.js, line 31 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3237#-KnuB59RP_52nDq65lsg-r1-31:-KnuJB_NZW24bxx9JXR3:bj42170) ([raw file](https://github.com/microsoft/chakracore/blob/cd79abbef5b7157181ea884af981da990bdaff7b/test/wasm.simd/replaceLaneTests.js#L31)):*
<details><summary><i>Previously, Krovatkin (Nick Korovaiko) wrote…</i></summary><blockquote>

we could change the indexing to a dict w/ an array view per each type where the type comes from a test name if this version is too ugly :-)
</blockquote></details>

Yeah this is a little hard to read and it feels hacky. I would just pass the typed array and the length by arguments.

---

*[test/wasm.simd/rlexe.xml, line 47 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3237#-KnuIZibbVIV8dc-qnSi:-KnuIZibbVIV8dc-qnSj:b-5z9iew) ([raw file](https://github.com/microsoft/chakracore/blob/cd79abbef5b7157181ea884af981da990bdaff7b/test/wasm.simd/rlexe.xml#L47)):*
> ```XML
> <test>
>   <default>
>     <files>splatNegTests.js</files>
> ```

This is not needed, rl will do a second run with prejit

---

*[test/wasm.simd/rlexe.xml, line 59 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3237#-KnuIdXaugwAn8XEmz44:-KnuIdXaugwAn8XEmz45:b-5z9iew) ([raw file](https://github.com/microsoft/chakracore/blob/cd79abbef5b7157181ea884af981da990bdaff7b/test/wasm.simd/rlexe.xml#L59)):*
> ```XML
> <test>
>   <default>
>     <files>replaceLaneTests.js</files>
> ```

This is not needed, rl will do a second run with prejit

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/3237)*
<!-- Sent from Reviewable.io -->

Krovatkin commented:
Review status: 10 of 15 files reviewed at latest revision, 5 unresolved discussions.

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 1286 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3237#-KnuCpaRG4KU5hoglIEA:-Knuhp-F0r_ZLSo4Z5dY:b-f7ism6) ([raw file](https://github.com/microsoft/chakracore/blob/cd79abbef5b7157181ea884af981da990bdaff7b/lib/WasmReader/WasmByteCodeGenerator.cpp#L1286)):*
<details><summary><i>Previously, Cellule (Michael Ferris) wrote…</i></summary><blockquote>

I think I would put the code in `CheckLaneIndex` in here that way if you register an index you have to check the index is not out of range.
Also, could we rename this function to something like `EmitLaneIndex(Js::OpCode requestingOp)`
</blockquote></details>

yup, looks much better that way

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 1303 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3237#-KnuDgEgunjxxElIemhi:-KnuhwHimxgnsRWk4FPt:b-7ql1td) ([raw file](https://github.com/microsoft/chakracore/blob/cd79abbef5b7157181ea884af981da990bdaff7b/lib/WasmReader/WasmByteCodeGenerator.cpp#L1303)):*
<details><summary><i>Previously, Cellule (Michael Ferris) wrote…</i></summary><blockquote>

use `PopEvalStack(WasmTypes::M128, _u("simd argument type mismatch"))` instead of checking the type manually after    
</blockquote></details>

fixed

---

*[test/wasm.simd/replaceLaneTests.js, line 31 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3237#-KnuB59RP_52nDq65lsg-r1-31:-Knuhyfe9MIWrBnMCE2_:buy4c3v) ([raw file](https://github.com/microsoft/chakracore/blob/cd79abbef5b7157181ea884af981da990bdaff7b/test/wasm.simd/replaceLaneTests.js#L31)):*
<details><summary><i>Previously, Cellule (Michael Ferris) wrote…</i></summary><blockquote>

Yeah this is a little hard to read and it feels hacky. I would just pass the typed array and the length by arguments.
</blockquote></details>

I tried passing both, the array view and length, and it seems like a lot of boilerplate that can be avoided. Also, it makes harder to add new tests since w/ two additional parameters it's easier to pass the wrong thing. 
I think the dict approach offers a nice compromise
* it's much easier to add new tests
* there's very little magic in helper functions to get the length and array view. 

https://github.com/Krovatkin/ChakraCore/blob/b998a530697ff2902e8666c6847f2798148ab8cc/test/wasm.simd/splatNegTests.js#L34
https://github.com/Krovatkin/ChakraCore/blob/b998a530697ff2902e8666c6847f2798148ab8cc/test/wasm.simd/splatNegTests.js#L47
https://github.com/Krovatkin/ChakraCore/blob/b998a530697ff2902e8666c6847f2798148ab8cc/test/wasm.simd/splatNegTests.js#L63

But if you insist on passing those as extra parameters, we could definitely do it. 

---

*[test/wasm.simd/rlexe.xml, line 47 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3237#-KnuIZibbVIV8dc-qnSi:-Knuiq6fxSRpCySzX0Le:b-7ql1td) ([raw file](https://github.com/microsoft/chakracore/blob/cd79abbef5b7157181ea884af981da990bdaff7b/test/wasm.simd/rlexe.xml#L47)):*
<details><summary><i>Previously, Cellule (Michael Ferris) wrote…</i></summary><blockquote>

This is not needed, rl will do a second run with prejit
</blockquote></details>

fixed

---

*[test/wasm.simd/rlexe.xml, line 59 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3237#-KnuIdXaugwAn8XEmz44:-Knuir9In8nUDl-AJfir:b-7ql1td) ([raw file](https://github.com/microsoft/chakracore/blob/cd79abbef5b7157181ea884af981da990bdaff7b/test/wasm.simd/rlexe.xml#L59)):*
<details><summary><i>Previously, Cellule (Michael Ferris) wrote…</i></summary><blockquote>

This is not needed, rl will do a second run with prejit
</blockquote></details>

fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/3237)*
<!-- Sent from Reviewable.io -->

Krovatkin commented:
@Cellule I was wondering if you could take another quick look? If everything looks good, I'll squash the commits and ask you to help me to merge this PR. @arunetm is still trying to figure out his perms issue.

Cellule commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[test/wasm.simd/replaceLaneTests.js, line 31 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3237#-KnuB59RP_52nDq65lsg-r1-31:-KnvCQM366APdT2e3Rss:b47469v) ([raw file](https://github.com/microsoft/chakracore/blob/cd79abbef5b7157181ea884af981da990bdaff7b/test/wasm.simd/replaceLaneTests.js#L31)):*
<details><summary><i>Previously, Krovatkin (Nick Korovaiko) wrote…</i></summary><blockquote>

I tried passing both, the array view and length, and it seems like a lot of boilerplate that can be avoided. Also, it makes harder to add new tests since w/ two additional parameters it's easier to pass the wrong thing. 
I think the dict approach offers a nice compromise
* it's much easier to add new tests
* there's very little magic in helper functions to get the length and array view. 

https://github.com/Krovatkin/ChakraCore/blob/b998a530697ff2902e8666c6847f2798148ab8cc/test/wasm.simd/splatNegTests.js#L34
https://github.com/Krovatkin/ChakraCore/blob/b998a530697ff2902e8666c6847f2798148ab8cc/test/wasm.simd/splatNegTests.js#L47
https://github.com/Krovatkin/ChakraCore/blob/b998a530697ff2902e8666c6847f2798148ab8cc/test/wasm.simd/splatNegTests.js#L63

But if you insist on passing those as extra parameters, we could definitely do it. 

</blockquote></details>

This looks pretty good!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/3237#-:-KnvCZweki88vqSf4vJ9:bnfp4nl)*
<!-- Sent from Reviewable.io -->

Cellule commented:
I'll try to merge it this afternoon

Krovatkin commented:
@Cellule Thank you very much!
We'll try to figure out perms issues on our end, so we at least don't have to burden you w/ merges.

Krovatkin commented:
@Cellule, awesome! Thanks a lot!
